python/ray/train/base_trainer.py

```diff
@@ -170,6 +170,22 @@ def __init__(
                 "#example-datasets-in-tune for more info."
             )
 
+    def __repr__(self):
+        non_default_arguments = []
+        if self.scaling_config != ScalingConfig():
+            non_default_arguments.append(f"scaling_config={self.scaling_config}")
+        if self.run_config != RunConfig():
+            non_default_arguments.append(f"run_config={self.run_config}")
+        if self.datasets != {}:
+            non_default_arguments.append(f"datasets={self.datasets}")
+        if self.preprocessor is not None:
+            non_default_arguments.append(f"preprocessor={self.preprocessor}")
+        if self.resume_from_checkpoint is not None:
+            non_default_arguments.append(
+                f"resume_from_checkpoint={self.resume_from_checkpoint}"
+            )
```

Comment:
can you do this a bit more programmatically? feel like this is very typo-prone.

ie:

```
arg_defaults = {
    "scaling_config": ScalingConfig(),
     "run_config": RunConfig(), 
     "datasets": ..., 
     "preprocessors": ..., 
     "resume_from_checkpoint": ...
}

for arg, default in arg_defaults.items():
    ...
    "{arg}={repr(getattr(self, arg))}
    ...
```

Reply:
Yeah. For context, I originally generated all of the docstrings progamatically (see [`ReprMixin` prototype](https://github.com/ray-project/ray/blob/1c9895b04be23f1ab7e39c13ecad000dec7d492c/python/ray/air/util/repr_mixin.py)), but decided to scrap that in favor of explicit `__repr__`s.

Reply:
Makes sense; Let's not do the mixin for sure.

Just do it within `def __repr__` but do it programmatically (it localizes the implementation to avoid too much indirection, and also makes it easier to avoid typos).